src/core/Akka/Actor/Futures.cs

```diff
@@ -117,92 +117,86 @@ public static Task<T> Ask<T>(this ICanTell self, object message, TimeSpan? timeo
-        /// Resolves <see cref="IActorRefProvider"/> for Ask pattern
-        /// </summary>
-        /// <param name="self">Reference to someone we are sending Ask request to</param>
-        /// <returns>Provider used for Ask pattern implementation</returns>
-        internal static IActorRefProvider ResolveProvider(ICanTell self)
-        {
-            if (self is ActorSelection)
-                return ResolveProvider(self.AsInstanceOf<ActorSelection>().Anchor);
-
-            if (self is IInternalActorRef)
-                return self.AsInstanceOf<IInternalActorRef>().Provider;
-            
-            if (ActorCell.Current != null)
-                return InternalCurrentActorCellKeeper.Current.SystemImpl.Provider;
-
-            return null;
-        }
-        
-        private static async Task<object> Ask(ICanTell self, Func<IActorRef, object> messageFactory, IActorRefProvider provider,
-            TimeSpan? timeout, CancellationToken cancellationToken)
-        {
-            TaskCompletionSource<object> result = TaskEx.NonBlockingTaskCompletionSource<object>();
+            var result = TaskEx.NonBlockingTaskCompletionSource<T>();
 
             CancellationTokenSource timeoutCancellation = null;
             timeout = timeout ?? provider.Settings.AskTimeout;
-            var ctrList = new List<CancellationTokenRegistration>(2);
+
+            CancellationTokenRegistration? ctr1 = null;
+            CancellationTokenRegistration? ctr2 = null;
+
 
             if (timeout != Timeout.InfiniteTimeSpan && timeout.Value > default(TimeSpan))
             {
                 timeoutCancellation = new CancellationTokenSource();
 
-                ctrList.Add(timeoutCancellation.Token.Register(() =>
+                ctr1 = timeoutCancellation.Token.Register(() =>
                 {
                     result.TrySetException(new AskTimeoutException($"Timeout after {timeout} seconds"));
-                }));
+                });
 
                 timeoutCancellation.CancelAfter(timeout.Value);
             }
 
             if (cancellationToken.CanBeCanceled)
             {
-                ctrList.Add(cancellationToken.Register(() => result.TrySetCanceled()));
+                ctr2 = cancellationToken.Register(() => result.TrySetCanceled());
             }
 
             //create a new tempcontainer path
             ActorPath path = provider.TempPath();
 
-            var future = new FutureActorRef(result, () => { }, path);
+            var future = new FutureActorRef<T>(result, () => { }, path);
             //The future actor needs to be registered in the temp container
             provider.RegisterTempActor(future, path);
             var message = messageFactory(future);
             self.Tell(message, future);
+            var prevContext = SynchronizationContext.Current;
 
+  
             try
             {
+                SynchronizationContext.SetSynchronizationContext(null);
                 return await result.Task;
```

Comment:
Can't get rid of this `await` unfortunately - it's much slower to do all of the cleanup using a `ContinueWith` than it is to just `await` the task and do cleanup in a `finally` block instead.